### DISCUSSION TITLE
Is it safe to publish my anon apikey and Authorization?

### DISCUSSION THREAD

**Discussion by michrome** [supabase/supabase#734] [Questions] [answered]
I have a function that I want to be available via anonymous [`fetch`](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API/Using_Fetch) calls **from the browser**, so I granted execution to anon with:

```sql
grant execute on function public.my_function_name() to anon;
```

I can now make calls like this:

```js
fetch("https://my-subdomain.supabase.co/rest/v1/rpc/my_function_name", {
  method: "POST",
  headers: {
    apikey:
      "my-anon-key",
    Authorization:
      "Bearer my-anon-bearer",
    "Content-Type": "application/json",
  },
  body: {
    key: "value",
  },
})
```

**Question: is it "safe" to include `my-anon-key` and `my-anon-bearer` in the source code of my web page?**

I know someone could try to use my anon key frequently enough for Supabase to rate limit/block it, but are there any other risks?

I know I could have an intermediate service (e.g. a server or a serverless function) to hide `my-anon-key` and `my-anon-bearer` so this question is really "do I have to?"

Thanks!

**Comment by kiwicopple** [comment, marked as answer]
Yes, it's safe *as long as you have Row Level Security turned on.* And with the proper security policies in place.

Here the Red padlock means that it is unsecure and the white means the table is secure

<img width="838" alt="image" src="https://user-images.githubusercontent.com/10214025/108309946-7e97d680-71ed-11eb-8088-6d151c1287f3.png">

**Reply by shorn** [reply]
The "RLS enabled" widget needs work to explain what's going on.  It's not self-evident that it's telling you the status of the row-level-security being enabled/disabled.

I'd suggest you should turn the thing into a toggle (perhaps inside a popup) that explicitly tells users that it represents "row level security enabled/disabled".

When RLS is enabled and there's no policies - the policy section should display some text explaining that because the user has enabled RLS, but not added policies, the table cannot be queried/altered.

Also, you probably need to start thinking about adding warnings about dangerous configurations.  For example: "RLS enabled but policy is set to `(true)` - anyone can read or write to this table".

**Comment by mstade** [comment]
> I know I could have an intermediate service (e.g. a server or a serverless function) to hide `my-anon-key` and `my-anon-bearer` so this question is really "do I have to?"

Further to this point, there may be other reasons why having a gateway in front makes sense. For example, we have some pretty specific auth provider requirements (see #544) and in order to implement this we would either:

1. Have to implement a custom provider that specifically implements the BankID flow for the interaction part of the auth flow – or, more generally – implement a provider that allows hooks to intercept the interaction part and the resume the flow with a pass/fail kind of response 
2. Put a gateway in front which lets us implement the custom interaction as part of the auth flow, and proxies all calls to the backend by generating JWTs forge backend to consume

We’re leaning toward the second option, not just because option #1 would mean we either have to deploy supabase ourselves – or – wait for such a provider to land in the product (we’d love to contribute, btw!) but also because as far as we can tell there are no options for key types (e.g. asymmetric keys) or management in supabase currently. Thus, we have to rely on the single secret (symmetric key) that supabase provides, and have no way of changing it short of nuking an environment and setting up a new one.

We’re currently looking at implementing this through a combination of AWS cognition and api gateway. Not sure if it’ll be successful, but if it is we’d be happy to contribute a tutorial on how to set this up.
